packages/celotool/src/e2e-tests/cip35_tests.ts

```diff
@@ -0,0 +1,454 @@
+const gatewayFeeRecipientAddress = '0xc77538d1e30C0e4ec44B0DcaD97FD3dc63fcaCC4'
+
+// Simple contract with a single constant
+const bytecode =
+  '0x608060405260008055348015601357600080fd5b5060358060216000396000f3006080604052600080fd00a165627a7a72305820c7f3f7c299940bb1d9b122d25e8f288817e45bbdeaccdd2f6e8801677ed934e70029'
+
+const verbose = false
+
+///////// Configurable values to run only some of the tests during development ////////////////
+// ReplayProtectionTests lets you skip or run only the replay-protection tests during dev
+// Value when committing should be "run"
+// tslint:disable-next-line
+let replayProtectionTests: 'run' | 'skip' | 'only' = 'run'
+// devFilter can be used during development to only run a subset of testcases.
+// But if you're going to commit you should set them all back to undefined (i.e. no filter).
+const devFilter: Filter = {
+  cipIsActivated: undefined,
+  lightNode: undefined,
+  ethCompatible: undefined,
+  contractCreation: undefined,
+  useFeeCurrency: undefined,
+  useGatewayFee: undefined,
+  useGatewayFeeRecipient: undefined,
+  sendRawTransaction: undefined,
+}
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+// Filter specifies which subset of cases to generate.
+// (e.g. {lightNode: true, sendRawTransaction: false} makes it only run cases which send through a light
+// node using `eth_sendRawTransaction`
+type Filter = Partial<TestCase>
+
+// TestCase describes the specific case we want to test
+interface TestCase {
+  cipIsActivated: boolean
+  lightNode: boolean
+  ethCompatible: boolean
+  contractCreation: boolean
+  useFeeCurrency: boolean
+  useGatewayFee: boolean
+  useGatewayFeeRecipient: boolean
+  sendRawTransaction: boolean // whether to use eth_sendRawTransaction ot eth_sendTransaction
+  errorString: string | null
+  errorReason: string | null
+}
+
+// generateTestCases is used to generate all the cases we want to test for a setup which
+// is either pre-Donut or post-Donut (cipIsActivated true means post-Donut)
+function generateTestCases(cipIsActivated: boolean) {
+  const cases: TestCase[] = []
+  if (devFilter.cipIsActivated !== undefined && devFilter.cipIsActivated !== cipIsActivated) {
+    // The devFilter is incompatible with the cipIsActivated value, so there are no cases to run
+    return cases
+  }
+  const getValues = (fieldFilter: boolean | undefined) => {
+    return fieldFilter === undefined ? [false, true] : [fieldFilter]
+  }
+  // Generate all possible combinations (but some are invalid and excluded using 'continue' below)
+  for (const lightNode of getValues(devFilter.lightNode)) {
+    for (const ethCompatible of getValues(devFilter.ethCompatible)) {
+      for (const contractCreation of getValues(devFilter.contractCreation)) {
+        for (const useFeeCurrency of getValues(devFilter.useFeeCurrency)) {
+          for (const useGatewayFee of getValues(devFilter.useGatewayFee)) {
```

Comment:
If this is true, the useGatewayFeeRecipient should also be true, right? Or am I missing some case?
if that's the case, we could collapse one for, or add an if as the one below ` if (sendRawTransaction && ethCompatible && hasCeloFields) {`

Reply:
It's possible to have `gatewayFee` non-zero and `gatewayFeeRecipient` nil.  See https://github.com/celo-org/celo-blockchain/blob/1a4f8f476dde6375397d9919e4557c4d32f87936/core/state_transition.go#L216

In reality, these two fields don't get much or any use, but I figured we might as well test all possible combinations.